File: src/Tree.ts
export class Directory {
    c_dir: Directory[] = [];
    c_file: File[] = [];
    parent: null | Directory = null;
    shown = false;
    level = 0;
    _mark = false;

    constructor(public time: Date, public name: string) {
        //
    }

    add_dir(d: Directory) {
        this.c_dir.push(d);
        d.level = this.level + 1;
        d.parent = this;
    }

    add_file(f: File) {
        this.c_file.push(f);
        f.level = this.level + 1;
        f.parent = this;
    }

    async add_n_dir(time: Date, p: string): Promise<any> {
        const parts = p.split("/");
        const name = parts.shift()!;
        const old = this.c_dir.find((c) => c.name === name);
        if (old && parts.length > 0) {
            return old.add_n_dir(time, parts.join("/"));
        }
        if (!old) {
            const ndir = new Directory(time, name);
            this.add_dir(ndir);
            if (parts.length > 0) {
                await ndir.add_n_dir(time, parts.join("/"));
            }
        }
    }
    find_dir(p: string): Directory | null {
        const parts = p.split("/");
        const name = parts.shift()!;
        const old = this.c_dir.find((c) => c.name === name);
        if (old && parts.length > 0) {
            return old.find_dir(parts.join("/"));
        } else if (old) {
            return old;
        } else {
            return null;
        }
    }
    async add_n_file(
        time: Date,
        p: string,
        fname: string,
        size: number,
        fullPath: string
    ) {
        if (p !== "") {
            await this.add_n_dir(time, p);
            const dir = this.find_dir(p) as Directory;
            const nfile = new File(time, fname, size, fullPath);
            dir.add_file(nfile);
        } else {
            const nfile = new File(time, fname, size, fullPath);
            this.add_file(nfile);
        }
    }

    //TODO: add sorting

    async process_to_stack(stack: (File | Directory)[], d: Directory) {
        for (let k = d.c_file.length - 1; k >= 0; k--) {
            stack.push(d.c_file[k]);
        }
        for (let k = d.c_dir.length - 1; k >= 0; k--) {
            stack.push(d.c_dir[k]);
        }
    }

    async to_linear() {
        const arr = [];
        const stack: (Directory | File)[] = [];
        await this.process_to_stack(stack, this);
        while (stack.length > 0) {
            const el = stack.pop();
            if (el instanceof File) {
                arr.push(el);
            } else if (el instanceof Directory) {
                arr.push(el);
                await this.process_to_stack(stack, el);
            }
        }
        return arr;
    }

    open() {
        this.c_dir.forEach((c) => (c.shown = true));
        this.c_file.forEach((c) => (c.shown = true));
    }
    close() {
        this.c_dir.forEach((c) => {
            c.shown = false;
            c.close();
        });
        this.c_file.forEach((c) => (c.shown = false));
    }
    toggle() {
        if (
            this.c_dir.find((c) => c.shown) ||
            this.c_file.find((c) => c.shown)
        ) {
            this.close();
        } else {
            this.open();
        }
    }

    mark() {
        this.c_dir.forEach((c) => c.mark());
        this.c_file.forEach((c) => c.mark());
    }
    unmark() {
        this.c_dir.forEach((c) => {
            c.unmark();
        });
        this.c_file.forEach((c) => c.unmark());
    }

    toggle_mark() {
        if (this._mark) {
            this.unmark();
            this._mark = false;
        } else {
            this.mark();
            this._mark = true;
        }
    }
    async process_to_stack_files(stack: (File | Directory)[], d: Directory) {
        for (let k = d.c_file.length - 1; k >= 0; k--) {
            if (!d.c_file[k].marked) continue;
            stack.push(d.c_file[k]);
        }
        for (let k = d.c_dir.length - 1; k >= 0; k--) {
            stack.push(d.c_dir[k]);
        }
    }

    async files_to_extract(): Promise<string[]> {
        const arr = [];
        const stack: (Directory | File)[] = [];
        await this.process_to_stack(stack, this);
        while (stack.length > 0) {
            const el = stack.pop();
            if (el instanceof File) {
                if (el.marked) arr.push(el.fullPath);
            } else if (el instanceof Directory) {
                await this.process_to_stack_files(stack, el);
            }
        }
        return arr;
    }
}

export class File {
    parent: null | Directory = null;
    shown = false;
    level = 0;
    marked = false;

    constructor(
        public time: Date,
        public name: string,
        public size: number,
        public fullPath: string
    ) {
        //
    }

    mark() {
        this.marked = true;
    }
    unmark() {
        this.marked = false;
    }
    toggle_mark() {
        this.marked = !this.marked;
    }
}

File: src/tree.rs
use std::collections::VecDeque;
use std::ops::Add;

#[derive(Debug)]
pub enum Node<'a> {
    Directory(&'a Directory),
    File(&'a File),
}

#[derive(Debug)]
pub struct Directory {
    c_dir: Vec<Directory>,
    c_file: Vec<File>,

    name: String,
    full_path: String,
    marked: bool,
    shown: bool,
    level: i32,
    child_shown: bool,
}

#[derive(Debug)]
pub struct File {
    name: String,
    full_path: String,
    marked: bool,
    shown: bool,
    level: i32,
}

pub trait FSNode {
    fn set_shown(&mut self, val: bool);
    fn get_level(&self) -> i32;
    fn is_marked(&self) -> bool;
    fn get_name(&self) -> &str;
    fn get_full_path(&self) -> &str;
    fn set_marked(&mut self, val: bool);
}

impl File {
    pub fn new(name: &str, full_path: &str, level: i32) -> Self {
        File {
            name: name.to_owned(),
            full_path: full_path.to_owned(),
            marked: false,
            shown: false,
            level,
        }
    }
}

impl FSNode for File {
    fn set_shown(&mut self, val: bool) {
        self.shown = val;
    }

    fn get_level(&self) -> i32 {
        self.level
    }

    fn is_marked(&self) -> bool {
        self.marked
    }

    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_full_path(&self) -> &str {
        &self.full_path
    }

    fn set_marked(&mut self, val: bool) {
        self.marked = val;
    }
}

impl Directory {
    pub fn new(name: &str, full_path: &str, level: i32) -> Self {
        Directory {
            c_dir: vec![],
            c_file: vec![],
            name: name.to_owned(),
            full_path: full_path.to_owned(),
            marked: false,
            shown: false,
            level,
            child_shown: false,
        }
    }

    //TODO: Add time and date
    //TODO: DOC: How it works how to use
    pub fn add_dir(&mut self, path: &str, f_path: &str) {
        let mut parts: VecDeque<&str> = path.split('/').collect();
        let c_name = parts
            .pop_front()
            .expect("Tree Creation Error, add_dir received broken path");
        let current_d = self.c_dir.iter_mut().find(|d| d.name == c_name);
        if let Some(current_d) = current_d {
            // Here parts length has to be longer than 0
            // the reason is creating files cant create directory
            // and directories cant be duplicate
            let mut n_path_str = String::with_capacity(path.len());
            parts.iter().enumerate().for_each(|(i, p)| {
                n_path_str.push_str(p);
                if i != parts.len() - 1 {
                    n_path_str.push_str("/");
                }
            });
            current_d.add_dir(&n_path_str, f_path);
        } else {
            /*if self.name == "$" {
                println!("Dir {} {} {}", c_name, f_path, path);
            }*/
            self.c_dir
                .push(Directory::new(c_name, f_path, self.level + 1));
        }
    }

    pub fn find_dir_mut(&mut self, path: &str) -> &mut Directory {
        let mut parts: VecDeque<&str> = path.split('/').collect();
        let c_name = parts
            .pop_front()
            .expect("Find Dir Error, find_dir_mut received broken path");
        let current_d = self
            .c_dir
            .iter_mut()
            .find(|d| d.name == c_name)
            .expect("Find Dir Error Path Does Not Exist");
        if parts.len() > 0 {
            let mut n_path_str = String::with_capacity(path.len());
            parts.iter().enumerate().for_each(|(i, p)| {
                n_path_str.push_str(p);
                if i != parts.len() - 1 {
                    n_path_str.push_str("/");
                }
            });
            current_d.find_dir_mut(&n_path_str)
        } else {
            current_d
        }
    }

    pub fn find_file_mut(&mut self, path: &str) -> &mut File {
        let last_slash = path.rfind("/");
        let (dir, filename) = match last_slash {
            Some(last_slash) => {
                let (dir_path, filename) = path.split_at(last_slash);
                let filename = &filename[1..filename.len()]; //because "/" stays with filename
                (self.find_dir_mut(dir_path), filename)
            }
            None => {
                // file at root
                (self, path)
            }
        };
        dir.c_file
            .iter_mut()
            .find(|el| el.get_name() == filename)
            .expect("Cant Find File Selected, Error In System")
    }

    //TODO: add time, date and size
    pub fn add_file(&mut self, path: &str) {
        let mut parts: VecDeque<&str> = path.split('/').collect();
        let c_name = parts
            .pop_back()
            .expect("Add File Error, add_file received broken path");
        if parts.len() > 0 {
            let mut n_path_str = String::with_capacity(path.len());
            parts.iter().enumerate().for_each(|(i, p)| {
                n_path_str.push_str(p);
                if i != parts.len() - 1 {
                    n_path_str.push_str("/");
                }
            });
            let dir = self.find_dir_mut(&n_path_str);
            dir.c_file.push(File::new(c_name, path, dir.level + 1));
        } else {
            self.c_file.push(File::new(c_name, path, self.level + 1));
        }
    }

    pub fn marked_list<'a>(&'a self) -> Vec<Node<'a>> {
        let mut v: Vec<Node<'a>> = vec![];
        for d in self.c_dir.iter() {
            if d.marked {
                v.push(Node::Directory(d));
            }
            let mut dir_res = d.marked_list();
            v.append(&mut dir_res);
        }
        for f in self.c_file.iter() {
            if f.marked {
                v.push(Node::File(f));
            }
        }
        v
    }

    pub fn shown_list<'a>(&'a self) -> Vec<Node<'a>> {
        let mut v: Vec<Node<'a>> = vec![];
        if self.shown {
            v.push(Node::Directory(self));
        }
        for d in self.c_dir.iter() {
            if d.shown {
                let mut dir_res = d.shown_list();
                v.append(&mut dir_res);
            }
        }
        for f in self.c_file.iter() {
            if f.shown {
                v.push(Node::File(f));
            }
        }
        v
    }

    pub fn set_shown_child(&mut self, val: bool) {
        for f in self.c_file.iter_mut() {
            f.set_shown(val);
        }
        for d in self.c_dir.iter_mut() {
            d.set_shown(val);
        }
    }

    pub fn set_shown_rec(&mut self, val: bool, initial: bool) {
        if !initial {
            self.set_shown(false);
        }
        for f in self.c_file.iter_mut() {
            f.set_shown(val);
        }
        for d in self.c_dir.iter_mut() {
            d.set_shown_rec(val, false);
        }
    }

    pub fn set_marked_rec(&mut self, val: bool) {
        self.set_marked(val);
        for f in self.c_file.iter_mut() {
            f.set_marked(val);
        }
        for d in self.c_dir.iter_mut() {
            d.set_marked_rec(val);
        }
    }

    pub fn open(&mut self, path: &str) {
        let dir = self.find_dir_mut(path);
        dir.set_shown_child(true);
    }

    pub fn close(&mut self, path: &str) {
        let dir = self.find_dir_mut(path);
        dir.set_shown_rec(false, true);
    }

    pub fn toggle_open(&mut self, path: &str) {
        let dir = if path == "$" {
            self
        } else {
            self.find_dir_mut(path)
        };
        if dir.child_shown {
            dir.set_shown_rec(false, true);
            dir.child_shown = false;
        } else {
            dir.set_shown_child(true);
            dir.child_shown = true;
        }
    }

    pub fn toggle_marked(&mut self, path: &str) {
        let dir = if path == "$" {
            self
        } else {
            self.find_dir_mut(path)
        };
        if dir.marked {
            dir.set_marked_rec(false);
        } else {
            dir.set_marked_rec(true);
        }
    }

    pub fn toggle_marked_file(&mut self, path: &str) {
        let f = self.find_file_mut(path);
        f.set_marked(!f.is_marked());
    }
}

impl FSNode for Directory {
    fn set_shown(&mut self, val: bool) {
        self.shown = val;
    }

    fn get_level(&self) -> i32 {
        self.level
    }

    fn is_marked(&self) -> bool {
        self.marked
    }

    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_full_path(&self) -> &str {
        &self.full_path
    }

    fn set_marked(&mut self, val: bool) {
        self.marked = val;
    }
}

impl<'a> FSNode for Node<'a> {
    fn get_full_path(&self) -> &str {
        match self {
            Node::File(v) => v.get_full_path(),
            Node::Directory(v) => v.get_full_path(),
        }
    }

    fn set_shown(&mut self, _: bool) {
        panic!("Unspported operation");
    }

    fn set_marked(&mut self, _: bool) {
        panic!("Unspported operation");
    }

    fn get_name(&self) -> &str {
        match self {
            Node::File(v) => v.get_name(),
            Node::Directory(v) => v.get_name(),
        }
    }

    fn is_marked(&self) -> bool {
        match self {
            Node::File(v) => v.is_marked(),
            Node::Directory(v) => v.is_marked(),
        }
    }
    fn get_level(&self) -> i32 {
        match self {
            Node::File(v) => v.get_level(),
            Node::Directory(v) => v.get_level(),
        }
    }
}
